File: push_swap_prototype/lab/push_swap/src/opti_sort_three.c
#include "models.h"

void	sort_three_top_a(t_ps *data, t_chunk *to_sort, t_stack *stk, int max);
void	sort_three_bottom_a(t_ps *data, t_chunk *to_sort, t_stack *stk, int max);
void	sort_three_top_b(t_ps *data, t_chunk *to_sort, t_stack *stk, int max);
void	sort_three_bottom_b(t_ps *data, t_chunk *to_sort, t_stack *stk, int max);
t_stack	*loc_to_stack(t_ps *data, t_loc loc);
int	chunk_max_value(t_ps *data, t_chunk *chunk);
void	swap_a(t_ps *data);
void	rotate_a(t_ps *data);
void	r_rotate_a(t_ps *data);
void	sort_two(t_ps *data, t_chunk *to_sort);
void	push_a(t_ps *data);
void	push_b(t_ps *data);
void	swap_b(t_ps *data);
void	r_rotate_b(t_ps *data);
int	next_down(t_stack *stk, int index);

void	sort_three(t_ps *data, t_chunk *to_sort)
{
	t_stack	*stk;
	int		max;

	stk = loc_to_stack(data, to_sort->loc);
	max = chunk_max_value(data, to_sort);
	if (to_sort->loc == TOP_A)
		sort_three_top_a(data, to_sort, stk, max);
	else if (to_sort->loc == BOTTOM_A)
		sort_three_bottom_a(data, to_sort, stk, max);
	else if (to_sort->loc == TOP_B)
		sort_three_top_b(data, to_sort, stk, max);
	else if (to_sort->loc == BOTTOM_B)
		sort_three_bottom_b(data, to_sort, stk, max);
}

void	sort_three_top_a(t_ps *data, t_chunk *to_sort, t_stack *stk, int max)
{
	if (stk->stack[stk->top] == max)
	{
		swap_a(data);
		rotate_a(data);
		swap_a(data);
		r_rotate_a(data);
	}
	else if (stk->stack[next_down(stk, stk->top)] == max)
	{
		rotate_a(data);
		swap_a(data);
		r_rotate_a(data);
	}
	to_sort->loc = TOP_A;
	to_sort->size -= 1;
	sort_two(data, to_sort);
}

void	sort_three_top_b(t_ps *data, t_chunk *to_sort, t_stack *stk, int max)
{
	push_a(data);
	if (stk->stack[stk->top] == max)
	{
		push_a(data);
		swap_a(data);
	}
	else if (stk->stack[next_down(stk, stk->top)] == max)
	{
		swap_b(data);
		push_a(data);
		swap_a(data);
	}
	else
		push_a(data);
	push_a(data);
	to_sort->loc = TOP_A;
	to_sort->size -= 1;
	sort_two(data, to_sort);
}

void	sort_three_bottom_a(t_ps *data, t_chunk *to_sort, t_stack *stk, int max)
{
	r_rotate_a(data);
	r_rotate_a(data);
	if (stk->stack[stk->top] == max)
	{
		swap_a(data);
		r_rotate_a(data);
	}
	else if (stk->stack[next_down(stk, stk->top)] == max)
		r_rotate_a(data);
	else
	{
		push_b(data);
		r_rotate_a(data);
		swap_a(data);
		push_a(data);
	}
	to_sort->loc = TOP_A;
	to_sort->size -= 1;
	sort_two(data, to_sort);
}

void	sort_three_bottom_b(t_ps *data, t_chunk *to_sort, t_stack *stk, int max)
{
	r_rotate_b(data);
	r_rotate_b(data);
	if (stk->stack[stk->top] == max)
	{
		push_a(data);
		r_rotate_b(data);
	}
	else if (stk->stack[next_down(stk, stk->top)] == max)
	{
		swap_b(data);
		push_a(data);
		r_rotate_b(data);
	}
	else
	{
		r_rotate_b(data);
		push_a(data);
	}
	to_sort->loc = TOP_B;
	to_sort->size -= 1;
	sort_two(data, to_sort);
}

File: push_swap_prototype/src/internal/solver/opti_sort_three.go
package solver

import (
	"push_swap_prototype/internal/chunk"
	"push_swap_prototype/internal/ops"
	"push_swap_prototype/internal/stack"
)

// sortThree třídí tři prvky v chunku shodně s C implementací
func SortThree(ps *ops.SortingState, chunk_item *chunk.Chunk) {
	
	
	stk := locToStack(ps, chunk_item.Loc)
	max := ChunkMaxValue(ps, chunk_item)


	switch chunk_item.Loc {
	case chunk.TOP_A:
		SortThreeTopA(ps, chunk_item, stk, max)
	case chunk.BOTTOM_A:
		SortThreeBottomA(ps, chunk_item, stk, max)
	case chunk.TOP_B:
		SortThreeTopB(ps, chunk_item, stk, max)
	case chunk.BOTTOM_B:
		SortThreeBottomB(ps, chunk_item, stk, max)
	}
}

func SortThreeTopA(ps *ops.SortingState, to_sort *chunk.Chunk, stk *stack.Stack, max int) {
	if stack.Value(stk, stk.Top) == max {
		ops.SwapA(ps)
		ops.RotateA(ps)
		ops.SwapA(ps)
		ops.ReverseRotateA(ps)
	} else if stack.Value(stk, stack.NextDown(stk, stk.Top)) == max {
		ops.RotateA(ps)
		ops.SwapA(ps)
		ops.ReverseRotateA(ps)
	}

	to_sort.Loc = chunk.TOP_A
	to_sort.Size -= 1
	
	SortTwo(ps, to_sort)
}

func SortThreeTopB(ps *ops.SortingState, to_sort *chunk.Chunk, stk *stack.Stack, max int) {
	ops.PushA(ps)

	if stack.Value(stk, stk.Top) == max {
		ops.PushA(ps)
		ops.SwapA(ps)
	} else if stack.Value(stk, stack.NextDown(stk, stk.Top)) == max {
		ops.SwapB(ps)
		ops.PushA(ps)
		ops.SwapA(ps)
	} else {
		ops.PushA(ps)
	}
	ops.PushA(ps)

	to_sort.Loc = chunk.TOP_A
	to_sort.Size -= 1

	SortTwo(ps, to_sort)
}

func SortThreeBottomA(ps *ops.SortingState, to_sort *chunk.Chunk, stk *stack.Stack, max int) {
	ops.ReverseRotateA(ps)
	ops.ReverseRotateA(ps)

	if stack.Value(stk, stk.Top) == max {
		ops.SwapA(ps)
		ops.ReverseRotateA(ps)
	} else if stack.Value(stk, stack.NextDown(stk, stk.Top)) == max {
		ops.ReverseRotateA(ps)
	}else {
		ops.PushB(ps)
		ops.ReverseRotateA(ps)
		ops.SwapA(ps)
		ops.PushA(ps)
	}

	to_sort.Loc = chunk.TOP_A
	to_sort.Size -= 1
	SortTwo(ps, to_sort)
}

func SortThreeBottomB(ps *ops.SortingState, to_sort *chunk.Chunk, stk *stack.Stack, max int) {
	ops.ReverseRotateB(ps)
	ops.ReverseRotateB(ps)

	if stack.Value(stk, stk.Top) == max {
		ops.PushA(ps)
		ops.ReverseRotateB(ps)
	}else if stack.Value(stk, stack.NextDown(stk, stk.Top)) == max {
		ops.SwapB(ps)
		ops.PushA(ps)
		ops.ReverseRotateB(ps)
	}	else {
		ops.ReverseRotateB(ps)
		ops.PushA(ps)
	}

	to_sort.Loc = chunk.TOP_B
	to_sort.Size -= 1
	SortTwo(ps, to_sort)
}